File: src/main/resources/static/js/comun/nombre-propio.js
/* =============================================================================
 *  nombre-propio.js
 *
 *  Pone mayuscula inicial en los campos de nombre y apellido al salir del campo.
 *  Se activa con data-nombre-propio en el input.
 *
 *  Por que. Lo que se guarda lo decide el servidor (NombrePropio.java), que
 *  aplica esta misma regla venga el dato de donde venga. Esto solo la muestra
 *  antes: quien tipeo "juan pablo" ve "Juan Pablo" al pasar al campo siguiente,
 *  en vez de enterarse en el listado de que el sistema lo cambio.
 *
 *  Al salir del campo y no mientras se escribe: corregir en cada tecla le mueve
 *  el cursor a quien esta escribiendo.
 *
 *  La regla tiene que ser la misma que la de NombrePropio.java. Si difieren, el
 *  campo muestra una cosa y se guarda otra.
 * ========================================================================== */

(function (window, document) {
    'use strict';

    // Las mismas de NombrePropio: van en minuscula, salvo cuando abren el campo.
    var PARTICULAS = ['de', 'del', 'la', 'las', 'los', 'y', 'e', 'da', 'das', 'do', 'dos', 'di', 'van', 'von'];

    var MAYUSCULA = /\p{Lu}/u;
    var MINUSCULA = /\p{Ll}/u;

    // Si el texto trae mayusculas y minusculas a la vez: alguien eligio como se escribe.
    function mezcla(t) {
        return MAYUSCULA.test(t) && MINUSCULA.test(t);
    }

    // Un tramo mezclado se respeta; si no, mayuscula inicial y el resto en minuscula.
    function tramo(t) {
        if (!t || mezcla(t)) return t;
        var m = t.toLowerCase();
        var inicial = String.fromCodePoint(m.codePointAt(0));
        return inicial.toUpperCase() + m.slice(inicial.length);
    }

    // Cada tramo entre guiones o apostrofos lleva su propia mayuscula: Perez-Gomez, O'Connor.
    // El split con parentesis conserva los separadores en las posiciones impares.
    function palabra(p, primera) {
        var minuscula = p.toLowerCase();
        if (!primera && !mezcla(p) && PARTICULAS.indexOf(minuscula) !== -1) return minuscula;
        return p.split(/(['’-])/).map(function (parte, i) {
            return i % 2 === 1 ? parte : tramo(parte);
        }).join('');
    }

    function normalizar(texto) {
        var limpio = texto.replace(/\s+/g, ' ').trim();
        if (!limpio) return limpio;
        return limpio.split(' ').map(function (p, i) {
            return palabra(p, i === 0);
        }).join(' ');
    }

    document.addEventListener('DOMContentLoaded', function () {
        document.querySelectorAll('input[data-nombre-propio]').forEach(function (campo) {
            // change y no blur: salta solo si el valor cambio, asi que pasar por el campo sin
            // tocarlo no reescribe el nombre que ya estaba guardado.
            campo.addEventListener('change', function () {
                var corregido = normalizar(campo.value);
                if (corregido !== campo.value) campo.value = corregido;
            });
        });
    });

    // Expuesto para poder probar la regla sin montar el formulario.
    window.NombrePropio = { normalizar: normalizar };

})(window, document);

File: src/main/java/edu/cent35/asistencias/model/NombrePropio.java
package edu.cent35.asistencias.model;

import java.util.Locale;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Escribe el nombre y el apellido de una persona con mayúscula inicial, se hayan tipeado en
 * minúscula o todo en mayúscula. Solo corrige lo que no tiene una forma elegida: una palabra
 * que ya mezcla mayúsculas y minúsculas, como McDonald o DiStefano, queda como se escribió.
 *
 * <p>El navegador aplica la misma regla al salir del campo ({@code js/comun/nombre-propio.js}),
 * pero solo para mostrar el resultado antes de guardar: lo que se guarda lo decide esta clase,
 * desde {@code DocenteService} y {@code UsuarioService}. Si se cambia una, se cambia la otra.
 */
public final class NombrePropio {

    // Las partículas de los nombres y apellidos compuestos: "María de los Ángeles", "González
    // de la Vega". Van en minúscula, salvo cuando abren el campo: un apellido que se escribe
    // solo empieza con mayúscula, "De la Fuente".
    private static final Set<String> PARTICULAS = Set.of(
        "de", "del", "la", "las", "los", "y", "e", "da", "das", "do", "dos", "di", "van", "von");

    // Con la clase Unicode, para que también cuente el espacio duro de un texto copiado.
    private static final Pattern ESPACIOS = Pattern.compile("\\s+", Pattern.UNICODE_CHARACTER_CLASS);

    private NombrePropio() {
    }

    // Devuelve el texto con mayúscula inicial en cada palabra y sin espacios de más.
    public static String normalizar(String texto) {
        if (texto == null) return null;
        String limpio = ESPACIOS.matcher(texto).replaceAll(" ").strip();
        if (limpio.isEmpty()) return limpio;

        String[] palabras = limpio.split(" ");
        StringBuilder sb = new StringBuilder(limpio.length());
        for (int i = 0; i < palabras.length; i++) {
            if (i > 0) sb.append(' ');
            sb.append(palabra(palabras[i], i == 0));
        }
        return sb.toString();
    }

    // Corrige una palabra. Cada tramo entre guiones o apóstrofos lleva su propia mayúscula:
    // Pérez-Gómez, O'Connor.
    private static String palabra(String p, boolean primera) {
        String minuscula = p.toLowerCase(Locale.ROOT);
        if (!primera && !mezcla(p) && PARTICULAS.contains(minuscula)) return minuscula;

        StringBuilder sb = new StringBuilder(p.length());
        int desde = 0;
        for (int i = 0; i < p.length(); i++) {
            char c = p.charAt(i);
            if (c == '-' || c == '\'' || c == '’') {
                sb.append(tramo(p.substring(desde, i))).append(c);
                desde = i + 1;
            }
        }
        return sb.append(tramo(p.substring(desde))).toString();
    }

    // Un tramo que mezcla mayúsculas y minúsculas se respeta; si no, lleva mayúscula inicial y
    // el resto en minúscula.
    private static String tramo(String t) {
        if (t.isEmpty() || mezcla(t)) return t;
        String m = t.toLowerCase(Locale.ROOT);
        int inicial = m.codePointAt(0);
        return new StringBuilder(m.length())
            .appendCodePoint(Character.toUpperCase(inicial))
            .append(m, Character.charCount(inicial), m.length())
            .toString();
    }

    // Si el texto trae mayúsculas y minúsculas a la vez: alguien eligió cómo se escribe.
    private static boolean mezcla(String t) {
        return t.codePoints().anyMatch(Character::isUpperCase)
            && t.codePoints().anyMatch(Character::isLowerCase);
    }
}
